### PR TITLE
zigbee2mqtt: 1.16.2 -> 1.22.1

### PR DESCRIPTION
**Pull request by sweber83**
###### Motivation for this change
Zigbee2mqtt still stuck @1.16.2

###### Things done
This PR builds on https://github.com/NixOS/nixpkgs/pull/147792 , adds a `postInstall` hook for typescript transpilation and then updates zigbee2mqtt to version 1.22.1.
It also fixes a small issue, where the module did still pass `dataDir` to the package.
Removes upgrade warning.

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/zigbee2mqtt/default.nix [hunk 1]
```diff
@@ -1,22 +1,24 @@
-{ pkgs, stdenv, dataDir ? "/opt/zigbee2mqtt/data", nixosTests }:
+{ pkgs, stdenv, nixosTests }:
 let
   package = (import ./node.nix { inherit pkgs; inherit (stdenv.hostPlatform) system; }).package;
 in
 package.override rec {
-  # don't upgrade! Newer versions cause stack overflows and fail trunk-combined
-  # see https://github.com/NixOS/nixpkgs/pull/118400
```
Comment (mweinelt) [inline pkgs/servers/zigbee2mqtt/default.nix:7]: > 1.1M	/nix/store/7nfnhncviphpls8yxix5d2xi2swa9hyi-zigbee2mqtt-1.22.1.drv

**Comment by markuskowa** [comment]
Duplicate of https://github.com/NixOS/nixpkgs/pull/147985

**Comment by mweinelt** [comment]
This is the continuation of the node2nix packaging for zigbee2mqtt. The other pr tries to rework it using yarn, which has issues that have been explained on the PR.

I prefer this changeset. Rolling out to my local setup as we speak.

**Comment by mweinelt** [comment]
@sweber83 Works for me. Added a minor security fixup to the systemd unit. PTAL.

**Review by mweinelt — APPROVED** [review]

**Comment by mweinelt** [comment]
Oh and commits touching the module should be prefixed with `nixos/zigbee2mqtt:`

File: pkgs/servers/zigbee2mqtt/update.sh [hunk 2]
```diff
@@ -1,30 +1,32 @@
...
-  --bypass-cache \
+node2nix \
+  --composition node.nix \
+  --lock npm-shrinkwrap.json \
+  --development \
   --no-copy-node-env \
-  --node-env ../../development/node-packages/node-env.nix
-rm package.json npm-shrinkwrap.json
+  --node-env ../../development/node-packages/node-env.nix \
+  --nodejs-14 \
```
Comment (mweinelt) [inline pkgs/servers/zigbee2mqtt/update.sh:24]: https://nodejs.org/en/about/releases/

Let's go with nodejs-16 if we can. Needs to generate the node-env at least.

Comment (sweber83) [reply pkgs/servers/zigbee2mqtt/update.sh:24]: If I run node2nix with the `--nodejs-16` flag, the node version in `node.nix` stays @ 12.
For node 14 it works as intended.

Comment (mweinelt) [reply pkgs/servers/zigbee2mqtt/update.sh:24]: I'll take that.

**Pull request merged**